### PR TITLE
Novacustom Wifi + Bluetooth switch

### PR DESCRIPTION
**Pull request by Mixss**

### CODE REVIEW

File: src/mainboard/clevo/tgl-u/romstage.c [hunk 1]
```diff
@@ -2,8 +2,41 @@
...
+    if (ret != CB_SUCCESS)
+            printk(BIOS_DEBUG, "Wifi + BT radios: failed to read curve selection from EFI vars, using board default\n");
+efi_err: ;
+#endif
+	if(!enable) {
+		uint32_t mask = memupd->FspmConfig.PcieRpEnableMask;
+		// disable PCIe root port 5
+		uint32_t disabled_root_port = ~((uint32_t)1 << 4);
+		memupd->FspmConfig.PcieRpEnableMask = mask & disabled_root_port;
+	}
```
Comment (miczyg1) [inline src/mainboard/clevo/tgl-u/romstage.c:40]: Please create a function from it.

File: src/mainboard/clevo/tgl-u/variants/nv40mz/romstage.c [hunk 2]
```diff
@@ -19,3 +19,5 @@ void variant_configure_fspm(FSPM_UPD *memupd)
 
 	memcfg_init(memupd, &board_cfg, &spd_info, half_populated);
 }
+
+
```
Comment (miczyg1) [inline src/mainboard/clevo/tgl-u/variants/nv40mz/romstage.c:23]: Please remove, extra new liens added

**Review by miczyg1 — CHANGES_REQUESTED** [review]
Please fix indentation ion all files, coreboot uses hard tabs 8 characters long. It seems your editor is set for 4 spaces, which is wrong.

File: src/mainboard/clevo/tgl-u/ramstage.c [hunk 3]
```diff
@@ -12,6 +12,7 @@
 #include <soc/ramstage.h>
 #include <variant/gpio.h>
 #include <variant/ramstage.h>
+#include <device/device.h>
```
Comment (miczyg1) [inline src/mainboard/clevo/tgl-u/ramstage.c:15]: This include is already present on line 3

File: src/mainboard/clevo/tgl-u/ramstage.c [hunk 4]
```diff
@@ -112,11 +113,42 @@ static void set_fan_curve(void)
...
+    ret = efi_fv_get_option(&rdev, &dasharo_system_features_guid, "EnableWifiBt",
+                            &enable, &size);
+
+    if (ret != CB_SUCCESS)
+            printk(BIOS_DEBUG, "Wifi + BT radios: failed to read curve selection from EFI vars, using board default\n");
+efi_err: ;
+#endif
+	struct device *wifi_card = pcidev_on_root(0x14, 3);
+	wifi_card->enabled = enable;
+	printk(BIOS_DEBUG, enable ? "raa Enabled\n" : "raa disabled\n");
```
Comment (miczyg1) [inline src/mainboard/clevo/tgl-u/ramstage.c:142]:
```suggestion
	printk(BIOS_DEBUG, enable ? "raa Enabled\n" : "raa disabled\n");
```

what is `raa`?

**Review by miczyg1 — CHANGES_REQUESTED** [review]

File: src/mainboard/clevo/tgl-u/variants/nv40mz/romstage.c (hunk 2 above)
Comment (Mixss) [reply src/mainboard/clevo/tgl-u/variants/nv40mz/romstage.c]: fixed in 31f28991bb1c1c1425e8a81e98819326d217a534

File: src/mainboard/clevo/tgl-u/ramstage.c (hunk 3 above)
Comment (Mixss) [reply src/mainboard/clevo/tgl-u/ramstage.c]: fixed in 31f28991bb1c1c1425e8a81e98819326d217a534

File: src/mainboard/clevo/tgl-u/ramstage.c (hunk 4 above)
Comment (Mixss) [reply src/mainboard/clevo/tgl-u/ramstage.c]: just debug print, removed in 31f28991bb1c1c1425e8a81e98819326d217a534

File: src/mainboard/clevo/tgl-u/romstage.c (hunk 1 above)
Comment (Mixss) [reply src/mainboard/clevo/tgl-u/romstage.c]: fixed in 5b544451a3f7956c7b56371134246418709d583d

File: src/mainboard/clevo/tgl-u/ramstage.c [hunk 5]
```diff
@@ -112,11 +112,41 @@ static void set_fan_curve(void)
...
+	};
+
+    if (smmstore_lookup_region(&rdev))
+		goto efi_err;
+
+    ret = efi_fv_get_option(&rdev, &dasharo_system_features_guid, "EnableWifiBt",
+                            &enable, &size);
+
+    if (ret != CB_SUCCESS)
+            printk(BIOS_DEBUG, "Wifi + BT radios: failed to read curve selection from EFI vars, using board default\n");
```
Comment (mkopec) [inline src/mainboard/clevo/tgl-u/ramstage.c:136]:
```suggestion
            printk(BIOS_DEBUG, "Wifi + BT radios: failed to read option selection from EFI vars, using board default\n");
```

File: src/mainboard/clevo/tgl-u/ramstage.c [hunk 6]
```diff
@@ -112,11 +112,41 @@ static void set_fan_curve(void)
...
+    uint32_t size = sizeof(enable);
+
+    const EFI_GUID dasharo_system_features_guid = {
+		0xd15b327e, 0xff2d, 0x4fc1, { 0xab, 0xf6, 0xc1, 0x2b, 0xd0, 0x8c, 0x13, 0x59 }
+	};
+
+    if (smmstore_lookup_region(&rdev))
+		goto efi_err;
+
+    ret = efi_fv_get_option(&rdev, &dasharo_system_features_guid, "EnableWifiBt",
```
Comment (miczyg1) [inline src/mainboard/clevo/tgl-u/ramstage.c:132]: Indentation is still not correct here

Comment (Mixss) [reply src/mainboard/clevo/tgl-u/ramstage.c]: fixed in 32991f56e6156a4c9daeb11f5e8be5882ac31396

**Comment by mkopec** [comment]
Moved adl-p option handling to mainboard_init in shared ramstage.c. Had to update relevant devicetree configs. Tested, enabling and disabling both work correctly on NS5x ADL with an AX201 card.

**Comment by mkopec** [comment]
Also did tgl-u and cleaned up remaining adl-p debug leftovers. Works on NV4x TGL with an AX200 card.

Should also test AX200 cards and test both cards on NS5x TGL as well.

File: src/mainboard/clevo/tgl-u/variants/ns50mu/ramstage.c [hunk 7]
```diff
@@ -12,4 +45,26 @@ void variant_configure_fsps(FSP_S_CONFIG *params)
...
+	radio_enable = get_wifi_bt_enable();
+
+	printk(BIOS_DEBUG, "Wireless is %sabled\n", radio_enable ? "en" : "dis");
+
+	wlan_dev->enabled = radio_enable;
+	cnvi_dev->enabled = radio_enable;
+	cfg->usb2_ports[9].enable = radio_enable;
+
+	cmd[0] = radio_enable;
+	system76_ec_smfi_cmd(CMD_WIFI_BT_ENABLEMENT_SET, 1, cmd);
```
Comment (miczyg1) [inline src/mainboard/clevo/tgl-u/variants/ns50mu/ramstage.c:69]:
```suggestion
	system76_ec_smfi_cmd(CMD_WIFI_BT_ENABLEMENT_SET, 1, (uint8_t *)&radio_enable);
```

File: src/mainboard/clevo/tgl-u/variants/nv40mz/ramstage.c [hunk 8]
```diff
@@ -43,3 +78,23 @@ static void mainboard_pre_device(void *unused) {
...
+	radio_enable = get_wifi_bt_enable();
+
+	printk(BIOS_DEBUG, "Wireless is %sabled\n", radio_enable ? "en" : "dis");
+
+	wlan_dev->enabled = radio_enable;
+	cnvi_dev->enabled = radio_enable;
+	cfg->usb2_ports[9].enable = radio_enable;
+
+	cmd[0] = radio_enable;
+	system76_ec_smfi_cmd(CMD_WIFI_BT_ENABLEMENT_SET, 1, cmd);
```
Comment (miczyg1) [inline src/mainboard/clevo/tgl-u/variants/nv40mz/ramstage.c:99]:
```suggestion
	system76_ec_smfi_cmd(CMD_WIFI_BT_ENABLEMENT_SET, 1, (uint8_t*)&radio_enable);
```

**Review by miczyg1 — CHANGES_REQUESTED** [review]

File: src/mainboard/clevo/adl-p/ramstage.c [hunk 9]
```diff
@@ -141,13 +169,31 @@ static uint8_t get_sleep_type_option(void)
...
+	printk(BIOS_DEBUG, "Wireless is %sabled\n", radio_enable ? "en" : "dis");
+
+	wlan_dev->enabled = radio_enable;
+	cnvi_dev->enabled = radio_enable;
+	cfg->cnvi_bt_core = radio_enable;
+	cfg->cnvi_bt_audio_offload = radio_enable;
+	cfg->usb2_ports[9].enable = radio_enable;
+
+	cmd[0] = radio_enable;
+	system76_ec_smfi_cmd(CMD_WIFI_BT_ENABLEMENT_SET, 1, cmd);
```
[the remark this answers is not in the document]
Comment (mkopec) [reply src/mainboard/clevo/adl-p/ramstage.c]: I guess we don't, suggestion applied

File: src/mainboard/clevo/tgl-u/variants/ns50mu/ramstage.c (hunk 7 above)
Comment (mkopec) [reply src/mainboard/clevo/tgl-u/variants/ns50mu/ramstage.c]: Done

File: src/mainboard/clevo/tgl-u/variants/nv40mz/ramstage.c (hunk 8 above)
Comment (mkopec) [reply src/mainboard/clevo/tgl-u/variants/nv40mz/ramstage.c]: Done

File: src/mainboard/clevo/adl-p/ramstage.c [hunk 10]
```diff
@@ -203,7 +249,7 @@ static void mainboard_smbios_strings(struct device *dev, struct smbios_type11 *t
 	if (result == 0) {
 		printk(BIOS_DEBUG, "EC firmware version: %s\n", ec_version);
 		t->count = smbios_add_string(t->eos, strconcat("EC firmware version: ",
-					     ec_version));
+						 ec_version));
```
[the remark this answers is not in the document]
Comment (mkopec) [reply src/mainboard/clevo/adl-p/ramstage.c]: My editor must've done something weird here, reverted.

**Comment by mkopec** [comment]
Tested with an AX210 (PCIe + USB on m.2 card), also works well.

**Comment by mkopec** [comment]
rebased to resolve conflicts

File: src/mainboard/clevo/adl-p/ramstage.c [hunk 11]
```diff
@@ -239,7 +283,7 @@ static void mainboard_smbios_strings(struct device *dev, struct smbios_type11 *t
 	if (result == 0) {
 		printk(BIOS_DEBUG, "EC firmware version: %s\n", ec_version);
 		t->count = smbios_add_string(t->eos, strconcat("EC firmware version: ",
-					     ec_version));
+							ec_version));
```
Comment (miczyg1) [inline src/mainboard/clevo/adl-p/ramstage.c:286]: Still wrong. Please revert it to how it was before the PR.

File: src/mainboard/clevo/adl-p/ramstage.c [hunk 12]
```diff
@@ -147,6 +147,34 @@ static void set_camera_enablement(void)
...
+
+	const EFI_GUID dasharo_system_features_guid = {
+		0xd15b327e, 0xff2d, 0x4fc1, { 0xab, 0xf6, 0xc1, 0x2b, 0xd0, 0x8c, 0x13, 0x59 }
+	};
+
+	if (smmstore_lookup_region(&rdev))
+		goto efi_err;
+
+	ret = efi_fv_get_option(&rdev, &dasharo_system_features_guid, "EnableWifiBt",
+							&enable, &size);
```
Comment (miczyg1) [inline src/mainboard/clevo/adl-p/ramstage.c:168]:
```suggestion
	ret = efi_fv_get_option(&rdev, &dasharo_system_features_guid, "EnableWifiBt",
				&enable, &size);
```

Please align all calls to efi_fv_get_option properly

**Review by miczyg1 — CHANGES_REQUESTED** [review]

File: src/mainboard/clevo/adl-p/ramstage.c (hunk 11 above)
Comment (mkopec) [reply src/mainboard/clevo/adl-p/ramstage.c:286]: fixed by ed5634a7b9e8fedd83bc84d662c4baa599bd4029

File: src/mainboard/clevo/adl-p/ramstage.c (hunk 12 above)
Comment (mkopec) [reply src/mainboard/clevo/adl-p/ramstage.c]: Aligned in d7f3d41ed9f54064d6c43eac17b77edd52f1a2cb

File: src/mainboard/clevo/adl-p/ramstage.c [hunk 13]
```diff
@@ -147,6 +147,34 @@ static void set_camera_enablement(void)
...
+	};
+
+	if (smmstore_lookup_region(&rdev))
+		goto efi_err;
+
+	ret = efi_fv_get_option(&rdev, &dasharo_system_features_guid, "EnableWifiBt",
+				&enable, &size);
+
+	if (ret != CB_SUCCESS)
+			printk(BIOS_DEBUG, "Wifi + BT radios: failed to read selection from EFI vars, using board default\n");
```
Comment (miczyg1) [inline src/mainboard/clevo/adl-p/ramstage.c:171]:
```suggestion
		printk(BIOS_DEBUG, "Wifi + BT radios: failed to read selection from EFI vars, using board default\n");
```

File: src/mainboard/clevo/tgl-u/variants/ns50mu/ramstage.c [hunk 14]
```diff
@@ -2,6 +2,39 @@
...
+	};
+
+	if (smmstore_lookup_region(&rdev))
+		goto efi_err;
+
+	ret = efi_fv_get_option(&rdev, &dasharo_system_features_guid, "EnableWifiBt",
+				&enable, &size);
+
+	if (ret != CB_SUCCESS)
+			printk(BIOS_DEBUG, "Wifi + BT radios: failed to read selection from EFI vars, using board default\n");
```
Comment (miczyg1) [inline src/mainboard/clevo/tgl-u/variants/ns50mu/ramstage.c:32]:
```suggestion
		printk(BIOS_DEBUG, "Wifi + BT radios: failed to read selection from EFI vars, using board default\n");
```

File: src/mainboard/clevo/tgl-u/variants/nv40mz/ramstage.c [hunk 15]
```diff
@@ -7,10 +7,43 @@
...
+	};
+
+	if (smmstore_lookup_region(&rdev))
+		goto efi_err;
+
+	ret = efi_fv_get_option(&rdev, &dasharo_system_features_guid, "EnableWifiBt",
+				&enable, &size);
+
+	if (ret != CB_SUCCESS)
+			printk(BIOS_DEBUG, "Wifi + BT radios: failed to read selection from EFI vars, using board default\n");
```
Comment (miczyg1) [inline src/mainboard/clevo/tgl-u/variants/nv40mz/ramstage.c:40]:
```suggestion
		printk(BIOS_DEBUG, "Wifi + BT radios: failed to read selection from EFI vars, using board default\n");
```

**Comment by miczyg1** [comment]
@mkopec would be great to have the edk2 revision bumped in configs too

**Comment by mkopec** [comment]
@miczyg1 Rebased, and bumped edk2 rev

**Review by miczyg1 — APPROVED** [review]

**Pull request merged**